File: compiler/python3/host/type_mapper.py
"""
Type mapping utilities for Python host compiler.
Converts MetaFFI types to Python type annotations and metaffi_type_info objects.
"""

from typing import Any, Optional, List, Tuple


def metaffi_type_to_python_type_annotation(metaffi_type: str, dimensions: int = 0, type_alias: Optional[str] = None) -> str:
    """
    Convert MetaFFI type string to Python type annotation.
    
    Args:
        metaffi_type: MetaFFI type string (e.g., "int32", "string8", "handle")
        dimensions: Number of array dimensions (0 = not an array)
        type_alias: Optional type alias for handle types
        
    Returns:
        Python type annotation string (e.g., "int", "str", "List[int]", "Any")
    """
    # Handle arrays
    if dimensions > 0:
        base_type = metaffi_type_to_python_type_annotation(metaffi_type, 0, type_alias)
        # For multi-dimensional arrays, nest List types
        result = base_type
        for _ in range(dimensions):
            result = f"List[{result}]"
        return result
    
    # Map base types
    type_mapping = {
        # Integers
        "int8": "int",
        "int16": "int",
        "int32": "int",
        "int64": "int",
        "uint8": "int",
        "uint16": "int",
        "uint32": "int",
        "uint64": "int",
        # Floats
        "float32": "float",
        "float64": "float",
        # Strings
        "string8": "str",
        "string16": "str",
        "string32": "str",
        # Characters
        "char8": "str",
        "char16": "str",
        "char32": "str",
        # Boolean
        "bool": "bool",
        # Handle
        "handle": type_alias if type_alias else "Any",
        # Other
        "any": "Any",
        "null": "None",
        "size": "int",
    }
    
    # Handle array types (e.g., "int32_array")
    if metaffi_type.endswith("_array"):
        base_type = metaffi_type[:-6]  # Remove "_array" suffix
        base_annotation = type_mapping.get(base_type, "Any")
        return f"List[{base_annotation}]"
    
    return type_mapping.get(metaffi_type, "Any")


def arg_to_metaffi_type_info(arg: "ArgDefinition", metaffi_types_module: Any) -> Any:
    """
    Convert ArgDefinition to metaffi_type_info object.
    
    Args:
        arg: ArgDefinition from IDL entities
        metaffi_types_module: metaffi_types module (required)
        
    Returns:
        metaffi_type_info object for use with load_entity()
    """
    if metaffi_types_module is None:
        raise ValueError("metaffi_types_module is required and cannot be None")
    
    metaffi_type_enum = _metaffi_type_string_to_enum(arg.type, arg.dimensions, metaffi_types_module)
    alias = arg.type_alias if arg.type_alias else None
    
    return metaffi_types_module.metaffi_type_info(
        metaffi_type=metaffi_type_enum,
        alias=alias,
        dims=arg.dimensions
    )


def _metaffi_type_string_to_enum(metaffi_type: str, dimensions: int, metaffi_types_module: Any):
    """
    Convert MetaFFI type string to MetaFFITypes enum.
    
    Args:
        metaffi_type: MetaFFI type string
        dimensions: Number of array dimensions
        metaffi_types_module: metaffi_types module (required)
        
    Returns:
        MetaFFITypes enum value
    """
    if metaffi_types_module is None:
        raise ValueError("metaffi_types_module is required and cannot be None")
    
    # Handle array types
    if dimensions > 0 or metaffi_type.endswith("_array"):
        base_type = metaffi_type.replace("_array", "")
        base_enum = _base_type_to_enum(base_type, metaffi_types_module)
        # Combine with array flag
        return base_enum | metaffi_types_module.MetaFFITypes.metaffi_array_type
    
    return _base_type_to_enum(metaffi_type, metaffi_types_module)


def _base_type_to_enum(metaffi_type: str, metaffi_types_module: Any):
    """
    Convert base MetaFFI type string to MetaFFITypes enum.
    
    Args:
        metaffi_type: MetaFFI type string
        metaffi_types_module: metaffi_types module (required)
    """
    if metaffi_types_module is None:
        raise ValueError("metaffi_types_module is required and cannot be None")
    MetaFFITypes = metaffi_types_module.MetaFFITypes
    mapping = {
        "int8": MetaFFITypes.metaffi_int8_type,
        "int16": MetaFFITypes.metaffi_int16_type,
        "int32": MetaFFITypes.metaffi_int32_type,
        "int64": MetaFFITypes.metaffi_int64_type,
        "uint8": MetaFFITypes.metaffi_uint8_type,
        "uint16": MetaFFITypes.metaffi_uint16_type,
        "uint32": MetaFFITypes.metaffi_uint32_type,
        "uint64": MetaFFITypes.metaffi_uint64_type,
        "float32": MetaFFITypes.metaffi_float32_type,
        "float64": MetaFFITypes.metaffi_float64_type,
        "bool": MetaFFITypes.metaffi_bool_type,
        "char8": MetaFFITypes.metaffi_char8_type,
        "char16": MetaFFITypes.metaffi_char16_type,
        "char32": MetaFFITypes.metaffi_char32_type,
        "string8": MetaFFITypes.metaffi_string8_type,
        "string16": MetaFFITypes.metaffi_string16_type,
        "string32": MetaFFITypes.metaffi_string32_type,
        "handle": MetaFFITypes.metaffi_handle_type,
        "any": MetaFFITypes.metaffi_any_type,
        "null": MetaFFITypes.metaffi_null_type,
        "size": MetaFFITypes.metaffi_size_type,
    }
    
    return mapping.get(metaffi_type, MetaFFITypes.metaffi_handle_type)


def parameters_to_metaffi_type_infos(parameters: List["ArgDefinition"], metaffi_types_module: Any) -> List:
    """
    Convert list of ArgDefinition parameters to list of metaffi_type_info objects.
    
    Args:
        parameters: List of ArgDefinition from function/method
        metaffi_types_module: metaffi_types module (required)
        
    Returns:
        List of metaffi_type_info objects
    """
    return [arg_to_metaffi_type_info(arg, metaffi_types_module) for arg in parameters]


def return_values_to_metaffi_type_infos(return_values: List["ArgDefinition"], metaffi_types_module: Any) -> List:
    """
    Convert list of ArgDefinition return values to list of metaffi_type_info objects.
    
    Args:
        return_values: List of ArgDefinition from function/method
        metaffi_types_module: metaffi_types module (required)
        
    Returns:
        List of metaffi_type_info objects
    """
    return [arg_to_metaffi_type_info(arg, metaffi_types_module) for arg in return_values]

File: compiler/cpp/host/type_mapper.cpp
#include "type_mapper.h"
#include <unordered_map>
#include <stdexcept>

namespace metaffi::compiler::cpp {

namespace {

// Static lookup tables for IDL base type → C++ scalar type and MetaFFI constant.
struct TypeEntry {
    const char* cpp_scalar;      // e.g. "int32_t"
    const char* metaffi_const;   // e.g. "metaffi_int32_type"
};

const std::unordered_map<std::string, TypeEntry> kTypeTable = {
    // --- Scalar types ---
    {"int8",    {"int8_t",         "metaffi_int8_type"}},
    {"int16",   {"int16_t",        "metaffi_int16_type"}},
    {"int32",   {"int32_t",        "metaffi_int32_type"}},
    {"int64",   {"int64_t",        "metaffi_int64_type"}},
    {"uint8",   {"uint8_t",        "metaffi_uint8_type"}},
    {"uint16",  {"uint16_t",       "metaffi_uint16_type"}},
    {"uint32",  {"uint32_t",       "metaffi_uint32_type"}},
    {"uint64",  {"uint64_t",       "metaffi_uint64_type"}},
    {"float32", {"float",          "metaffi_float32_type"}},
    {"float64", {"double",         "metaffi_float64_type"}},
    {"bool",    {"bool",           "metaffi_bool_type"}},
    {"string8", {"std::string",    "metaffi_string8_type"}},
    {"string16",{"std::u16string", "metaffi_string16_type"}},
    {"string32",{"std::u32string", "metaffi_string32_type"}},
    {"char8",   {"char",           "metaffi_char8_type"}},
    {"char16",  {"char16_t",       "metaffi_char16_type"}},
    {"char32",  {"char32_t",       "metaffi_char32_type"}},
    {"handle",  {"metaffi_handle", "metaffi_handle_type"}},
    {"size",    {"uint64_t",       "metaffi_size_type"}},
    {"any",     {"void*",          "metaffi_any_type"}},
    {"null",    {"void*",          "metaffi_null_type"}},

    {"callable",{"cdt_metaffi_callable*", "metaffi_callable_type"}},

    // --- Non-packed array types (idl type = base + "_array") ---
    {"int8_array",    {"int8_t",         "metaffi_int8_array_type"}},
    {"int16_array",   {"int16_t",        "metaffi_int16_array_type"}},
    {"int32_array",   {"int32_t",        "metaffi_int32_array_type"}},
    {"int64_array",   {"int64_t",        "metaffi_int64_array_type"}},
    {"uint8_array",   {"uint8_t",        "metaffi_uint8_array_type"}},
    {"uint16_array",  {"uint16_t",       "metaffi_uint16_array_type"}},
    {"uint32_array",  {"uint32_t",       "metaffi_uint32_array_type"}},
    {"uint64_array",  {"uint64_t",       "metaffi_uint64_array_type"}},
    {"float32_array", {"float",          "metaffi_float32_array_type"}},
    {"float64_array", {"double",         "metaffi_float64_array_type"}},
    {"bool_array",    {"bool",           "metaffi_bool_array_type"}},
    {"string8_array", {"std::string",    "metaffi_string8_array_type"}},
    {"char8_array",   {"char",           "metaffi_char8_array_type"}},
    {"char16_array",  {"char16_t",       "metaffi_char16_array_type"}},
    {"char32_array",  {"char32_t",       "metaffi_char32_array_type"}},
    {"handle_array",  {"metaffi_handle", "metaffi_handle_array_type"}},
    {"size_array",    {"uint64_t",       "metaffi_size_array_type"}},

    // --- Packed array base types (strip_array_suffix leaves these after removing "_array") ---
    {"int8_packed",    {"int8_t",         "metaffi_int8_packed_array_type"}},
    {"int16_packed",   {"int16_t",        "metaffi_int16_packed_array_type"}},
    {"int32_packed",   {"int32_t",        "metaffi_int32_packed_array_type"}},
    {"int64_packed",   {"int64_t",        "metaffi_int64_packed_array_type"}},
    {"uint8_packed",   {"uint8_t",        "metaffi_uint8_packed_array_type"}},
    {"uint16_packed",  {"uint16_t",       "metaffi_uint16_packed_array_type"}},
    {"uint32_packed",  {"uint32_t",       "metaffi_uint32_packed_array_type"}},
    {"uint64_packed",  {"uint64_t",       "metaffi_uint64_packed_array_type"}},
    {"float32_packed", {"float",          "metaffi_float32_packed_array_type"}},
    {"float64_packed", {"double",         "metaffi_float64_packed_array_type"}},
    {"bool_packed",    {"bool",           "metaffi_bool_packed_array_type"}},
    {"string8_packed", {"std::string",    "metaffi_string8_packed_array_type"}},
    {"char8_packed",   {"char",           "metaffi_char8_packed_array_type"}},
    {"char16_packed",  {"char16_t",       "metaffi_char16_packed_array_type"}},
    {"char32_packed",  {"char32_t",       "metaffi_char32_packed_array_type"}},
    {"handle_packed",  {"metaffi_handle", "metaffi_handle_packed_array_type"}},
    {"size_packed",    {"uint64_t",       "metaffi_size_packed_array_type"}},
};

} // namespace

// ---------------------------------------------------------------------------
// Private helpers
// ---------------------------------------------------------------------------

std::string CppTypeMapper::strip_array_suffix(const std::string& idl_type) {
    const std::string suffix = "_array";
    if (idl_type.size() > suffix.size() &&
        idl_type.compare(idl_type.size() - suffix.size(), suffix.size(), suffix) == 0) {
        return idl_type.substr(0, idl_type.size() - suffix.size());
    }
    return idl_type;
}

std::string CppTypeMapper::scalar_cpp_type(const std::string& base_type) {
    auto it = kTypeTable.find(base_type);
    if (it == kTypeTable.end()) {
        throw std::runtime_error("CppTypeMapper: unknown MetaFFI type '" + base_type + "'");
    }
    return it->second.cpp_scalar;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

std::string CppTypeMapper::cpp_type(const std::string& idl_type, int dimensions) {
    // Strip _array suffix to get the base type
    const std::string base = strip_array_suffix(idl_type);
    const std::string scalar = scalar_cpp_type(base);

    // An _array suffix with dim=0 implies at least one dimension of wrapping
    const bool has_array_suffix = (base != idl_type);
    const int effective_dims = (has_array_suffix && dimensions == 0) ? 1 : dimensions;

    // Wrap in std::vector<> for each dimension
    if (effective_dims > 0) {
        std::string result = scalar;
        for (int d = 0; d < effective_dims; ++d) {
            result = "std::vector<" + result + ">";
        }
        return result;
    }

    return scalar;
}

std::string CppTypeMapper::metaffi_constant(const std::string& idl_type) {
    // Try the full type first (handles int64_array, int64_packed_array, etc.)
    {
        auto it = kTypeTable.find(idl_type);
        if (it != kTypeTable.end()) {
            return it->second.metaffi_const;
        }
    }

    // Fall back: strip _array suffix and look up the base scalar/packed entry
    const std::string base = strip_array_suffix(idl_type);
    auto it = kTypeTable.find(base);
    if (it == kTypeTable.end()) {
        throw std::runtime_error("CppTypeMapper: unknown MetaFFI type '" + idl_type + "'");
    }

    return it->second.metaffi_const;
}

std::string CppTypeMapper::type_info_literal(const std::string& idl_type, int dimensions) {
    // Determine effective dimensions: IDL dimension field or _array suffix implies dim ≥ 1
    const std::string base = strip_array_suffix(idl_type);
    bool has_array_suffix = (base != idl_type);

    int effective_dims = dimensions;
    if (has_array_suffix && effective_dims == 0) {
        effective_dims = 1;
    }

    const std::string constant = metaffi_constant(idl_type);

    // Use explicit constructor calls to avoid brace-init ambiguity:
    // {type, N} in a vector initializer creates N+1 elements instead of one.
    if (effective_dims > 0) {
        return "metaffi_type_info(" + constant + ", nullptr, false, " + std::to_string(effective_dims) + ")";
    }
    return "metaffi_type_info(" + constant + ")";
}

} // namespace metaffi::compiler::cpp
